server/src/main/java/io/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,327 @@
+import io.druid.discovery.DiscoveryDruidNode;
+import io.druid.discovery.DruidNodeDiscovery;
+import io.druid.discovery.DruidNodeDiscoveryProvider;
+import io.druid.guice.ManageLifecycle;
+import io.druid.guice.annotations.Json;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.lifecycle.LifecycleStart;
+import io.druid.java.util.common.lifecycle.LifecycleStop;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.server.initialization.CuratorDiscoveryConfig;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.cache.PathChildrenCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheEvent;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheListener;
+import org.apache.curator.utils.ZKPaths;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+
+/**
+ */
+@ManageLifecycle
+public class CuratorDruidNodeDiscoveryProvider extends DruidNodeDiscoveryProvider
+{
+  private static final Logger log = new Logger(CuratorDruidNodeDiscoveryProvider.class);
+
+  private final CuratorFramework curatorFramework;
+  private final CuratorDiscoveryConfig config;
+  private final ObjectMapper jsonMapper;
+
+  private final ExecutorService listenerExecutor;
+
+  private final Map<String, NodeTypeWatcher> nodeTypeWatchers = new ConcurrentHashMap<>();
+
+  private boolean stopped = false;
+
+  @Inject
+  public CuratorDruidNodeDiscoveryProvider(
+      CuratorFramework curatorFramework,
+      CuratorDiscoveryConfig config,
+      @Json ObjectMapper jsonMapper
+  )
+  {
+    // This is single-threaded to ensure that all listener calls are executed precisesly in the oder of add/remove
+    // event occurences.
+    this.listenerExecutor = Execs.singleThreaded("CuratorDruidNodeDiscoveryProvider-ListenerExecutor");
+    this.curatorFramework = curatorFramework;
+    this.config = config;
+    this.jsonMapper = jsonMapper;
+  }
+
+
+
+  @Override
+  public synchronized DruidNodeDiscovery getForNodeType(String nodeType)
+  {
+    if (stopped) {
+      throw new ISE("Provider has been stopped.");
+    }
+
+    NodeTypeWatcher nodeTypeWatcher = nodeTypeWatchers.get(nodeType);
+    if (nodeTypeWatcher == null) {
+      log.info("Creating/Starting NodeTypeWatcher for nodeType [%s].", nodeType);
+      nodeTypeWatcher = new NodeTypeWatcher(
+          listenerExecutor,
+          curatorFramework,
+          config.getInternalDiscoveryPath(),
+          jsonMapper,
+          nodeType
+      );
+      nodeTypeWatcher.start();
+      nodeTypeWatchers.put(nodeType, nodeTypeWatcher);
+      log.info("Created/Started NodeTypeWatcher for nodeType [%s].", nodeType);
+    }
+
+    return nodeTypeWatchers.get(nodeType);
+  }
+
+  @LifecycleStart
+  public synchronized void start()
+  {
+    log.info("started");
```

Comment:
changed to create listeners executor in start instead of constructor, so now it does something.